h/static/scripts/cross-frame-sync.coffee

```diff
@@ -0,0 +1,197 @@
+
+    # Register remotely invokable methods
+    for method, func in @channelListeners
+      bridge.on(method, func)
+
+  _sync_cache: (channel) =>
+    # Synchronise (here to there) the items in our cache
+    annotations = (this._format a for t, a of @cache)
+    if annotations.length
+      channel.notify
+        method: 'loadAnnotations'
+        params: annotations
+
+  _emit: (args...) =>
+    @$rootScope.$emit.call(@$rootScope, args...)
+
+  _on: (event, handler) =>
+    @$rootScope.$on(event, handler)
+
+  getAnnotationForTag: (tag) ->
+    @cache[tag] or null
+
+  # Handlers for events coming from this frame, to send them across the channel
+  eventListeners:
+    'beforeAnnotationCreated': @beforeAnnotationCreated
+    'annotationCreated': @annotationCreated
+    'annotationUpdated': @annotationUpdated
+    'annotationDeleted': @annotationDeleted
+    'annotationsLoaded': @annotationsLoaded
+
+
+  _mkCallRemotelyAndParseResults: (method, callBack) ->
+    fn = (annotation) ->
+      # Wrap the callback function to first parse returned items
+      wrappedCallback = (failure, results) =>
+        unless failure?
+          this._parseResults results
+        callBack? failure, results
+
+      # Call the remote method
+      options =
+        method: method
+        callback: wrappedCallback
+        params: this._format(annotation)
+      @bridge.call(options)
+
+  # Handlers for sidebar events on $rootScope. Pass events through channel to guest code.
+
+  beforeAnnotationCreated: (event, annotation) =>
+    return if annotation.$$tag?
+    this._mkCallRemotelyAndParseResults('beforeCreateAnnotation')(annotation)
+    this
+
+  annotationCreated: (event, annotation) =>
+    return unless annotation.$$tag? and @cache[annotation.$$tag]
+    this._mkCallRemotelyAndParseResults('createAnnotation')(annotation)
+    this
+
+  annotationUpdated: (event, annotation) =>
+    return unless annotation.$$tag? and @cache[annotation.$$tag]
+    this._mkCallRemotelyAndParseResults('updateAnnotation')(annotation)
+    this
+
+  annotationDeleted: (event, annotation) =>
+    return unless annotation.$$tag? and @cache[annotation.$$tag]
+    onFailure = (err) =>
+      if err then @annotator.setupAnnotation annotation # TODO
+      else delete @cache[annotation.$$tag]
+    this._mkCallRemotelyAndParseResults('deleteAnnotation', onFailure)(annotation)
+    this
+
+  annotationsLoaded: (event, annotations) =>
+    annotations = (this._format a for a in annotations when not a.$$tag)
+    return unless annotations.length
+    this._notify
+      method: 'loadAnnotations'
+      params: annotations
+    this
+
+  # Parse returned annotations to update cache with any changes made remotely
+  _parseResults: (results) ->
+    if Array.isArray(results[0])
+      acc = []
+      foldFn = (_, cur) =>
+        (this._parse(a) for a in cur)
+    else
+      acc = {}
+      foldFn = (_, cur) =>
+        this._parse(cur)
+    results.reduce(foldFn, acc)
```

Comment:
@Treora I'm not sure this works as expected, this will result in only the last passed result being returned. I would expect `.map` to be used here.

```coffee
results.map(this._parse, this)
```